sqlite_gis.c

```diff
@@ -0,0 +1,273 @@
+/*-------------------------------------------------------------------------
+ *
+ * SQLite Foreign Data Wrapper for PostgreSQL
+ *
+ * Portions Copyright (c) 2018, TOSHIBA CORPORATION
+ *
+ * IDENTIFICATION
+ * 		sqlite_gis.c
+ *
+ * Routines that convert between SpatiaLite BLOB storage form and PostGIS EWKB
+ * and some functions about detecting data type names from PostGIS set
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+#include "sqlite_fdw.h"
+
+#ifdef SQLITE_FDW_GIS_ENABLE
+#include <spatialite.h>
+#endif
+
+#include <sqlite3.h>
+
+#include "nodes/makefuncs.h"
+#include "parser/parse_type.h"
+#include "utils/lsyscache.h"
+
+#ifdef SQLITE_FDW_GIS_ENABLE
+static void
+			common_EWKB_error (Form_pg_attribute att, int len, const char* data, bool direction_to_pg);
+static char*
+			getHexFormOfBlob(blobOutput b);
+#endif
+
+#define		EWKT_SRID_TEST_PREFIX_LEN 12
+
+/*
+ * This is data types from PostGIS 3.4. For a newer version use the following query
+ * select pt.typname, pt."oid", pd.description,
+ *		  pt.typinput, pt.typoutput
+ *   from pg_catalog.pg_type pt
+ *  inner join pg_catalog.pg_description pd
+ *	 on pt."oid" = pd.objoid
+ *  where pd.description like 'postgis %'
+ */
+const char *postGisSpecificTypes[] =
+{
+	"addbandarg", "box2d", "box3d", "geometry_dump", "geomval",
+	"getfaceedges_returntype", "rastbandarg", "raster", "reclassarg",
+	"summarystats", "topoelement", "topoelementarray", "topogeometry",
+	"unionarg", "validatetopology_returntype", NULL
+};
+
+const char *postGisSQLiteCompatibleTypes[] = { "geometry", "geography", NULL };
+
+#ifdef SQLITE_FDW_GIS_ENABLE
+/*
+ * SpatiaLiteAsPostGISgeom:
+ * Gives SpatiaLite BLOB, returns hex value string for PostGIS/GEOS input function
+ */
+char *
+SpatiaLiteAsPostGISgeom (blobOutput spatiaLiteBlob, Form_pg_attribute att)
+{
+	gaiaOutBuffer out_buf;
+	gaiaGeomCollPtr geo = NULL;
+	int gpkg_amphibious = 0;
+	int gpkg_mode = 0;
+	int res_len = 0;
+	char *res = NULL;
+
+	geo = gaiaFromSpatiaLiteBlobWkbEx ((unsigned char *)spatiaLiteBlob.dat,
+										spatiaLiteBlob.len,
+										gpkg_mode,
+										gpkg_amphibious);
+
+	if (!geo)
+	{
+		common_EWKB_error (att,
+						   spatiaLiteBlob.len,
+						   getHexFormOfBlob(spatiaLiteBlob),
+						   true);
+	}
+
+	gaiaOutBufferInitialize (&out_buf);
+	gaiaToEWKB (&out_buf, geo);
+	if (out_buf.Error || out_buf.Buffer == NULL)
+	{
+		gaiaOutBufferReset (&out_buf);
+		common_EWKB_error (att,
+						   spatiaLiteBlob.len,
+						   getHexFormOfBlob(spatiaLiteBlob),
+						   true);
+	}
+
+	res_len = strlen(out_buf.Buffer);
+	res = (char*) palloc(res_len*sizeof(char));
+	strcpy(res, out_buf.Buffer);
```

Comment:
@mkgrgis The reason for `WARNING:  problem in alloc set ExprContext: detected write past chunk end in block 0x39bfb10, chunk 0x39c03a8` is here.  
`strlen` does not count the terminating null character. It means, your `palloc` does not allocate for terminating null character.  
However, `strcpy` copies data including terminating null character. It means you copied terminating null character into an unallocated memory. It causes the WARNING.  
Please increase `res_len` by 1 to fix it.

Reply:
Fixed in https://github.com/pgspider/sqlite_fdw/pull/96/commits/17c12f88acb8167a2a496ce99c070d2f4cd79e15, @son-phamngoc ! Thanks! Unfortunately I have no experience with memory context checking or `-fno-omit-frame-pointer` option. Sometimes this is very useful to check some implementations against wrong memory allocating, but I can not :-(

Reply:
Confirmed. Thanks. Please help me to resolve this comment.